src/main/java/com/synopsys/integration/detect/configuration/validation/DetectConfigurationValidator.java

```diff
@@ -0,0 +1,123 @@
+            PropertyDeprecationInfo deprecationInfo = property.getPropertyDeprecationInfo();
+            String deprecationMessage = deprecationInfo.getDeprecationText();
+            String propertyKey = property.getKey();
+
+            additionalNotes.put(propertyKey, "\t *** DEPRECATED ***");
+
+            deprecationMessages.put(propertyKey, Collections.singletonList(deprecationMessage));
+            DetectIssue.publish(eventSystem, DetectIssueType.DEPRECATION, propertyKey, "\t" + deprecationMessage);
+
+            if (!hasUsedFailureProperties && detectInfo.getDetectMajorVersion() >= deprecationInfo.getFailInVersion().getIntValue()) {
+                hasUsedFailureProperties = true;
+            }
+        }
+
+        return new DetectConfigurationState(detectConfiguration, additionalNotes, deprecationMessages, hasUsedFailureProperties);
+    }
+
+    public void printConfiguration(PropertyConfigurationHelpContext detectConfigurationReporter, Map<String, String> additionalNotes) throws IllegalAccessException {
+        detectConfigurationReporter.printCurrentValues(infoLogReportWriter::writeLine, DetectProperties.allProperties(), additionalNotes);
+    }
+
+
+    public Optional<DetectBootResult> validateConfiguration(PropertyConfigurationHelpContext detectConfigurationReporter, PropertyConfiguration detectConfiguration, Map<String, List<String>> deprecationMessages, boolean noFailurePropertiesUsed) throws IllegalAccessException {
+        //Check for options that are just plain bad, ie giving an detector type we don't know about.
+        Map<String, List<String>> errorMap = detectConfigurationReporter.findPropertyParseErrors(DetectProperties.allProperties());
+        if (errorMap.size() > 0) {
+            Map.Entry<String, List<String>> entry = errorMap.entrySet().iterator().next();
+            return Optional.of(DetectBootResult.exception(new DetectUserFriendlyException(entry.getKey() + ": " + entry.getValue().get(0), ExitCodeType.FAILURE_GENERAL_ERROR), detectConfiguration));
+        }
+
+        if (noFailurePropertiesUsed) {
+            detectConfigurationReporter.printPropertyErrors(infoLogReportWriter::writeLine, DetectProperties.allProperties(), deprecationMessages);
+
+            logger.warn(StringUtils.repeat("=", 60));
+            logger.warn("Configuration is using deprecated properties that must be updated for this major version.");
+            logger.warn("You MUST fix these deprecation issues for detect to proceed.");
+            logger.warn(String.format("To ignore these messages and force detect to exit with success supply --%s=true", DetectProperties.DETECT_FORCE_SUCCESS.getProperty().getKey()));
+            logger.warn("This will not force detect to run, but it will pretend to have succeeded.");
+            logger.warn(StringUtils.repeat("=", 60));
+
+            eventSystem.publishEvent(Event.ExitCode, new ExitCodeRequest(ExitCodeType.FAILURE_CONFIGURATION));
```

Comment:
I'd really prefer this simply return some object that describes what it did. ConfigurationValidationResult for example, and the DetectBoot piece checks for ".hasFailureProperty" or something which triggers the event and converts the validation result to a boot result.

Reply:
Sounds good to me, I'll take a crack at that. I agree, it's a bit strange that the method signature returns an Optional<DetectBootResult>